File: 30-Day LeetCoding Challenge/Python/leftMostColumnWithOne.py
# """
# This is BinaryMatrix's API interface.
# You should not implement it, or speculate about its implementation
# """
#class BinaryMatrix(object):
#    def get(self, row, col):
#        """
#        :type row : int, col : int
#        :rtype int
#        """
#
#    def dimensions:
#        """
#        :rtype list[]
#        """

class Solution(object):
    def leftMostColumnWithOne(self, binaryMatrix):
        """
        :type binaryMatrix: BinaryMatrix
        :rtype: int
        """
        dimensions = binaryMatrix.dimensions()
        row = dimensions[0] - 1
        col = dimensions[1] - 1
        hasOne = False
        while (row > -1) and (col > -1):
            value = binaryMatrix.get(row, col)
            if value == 0:
                row -= 1
            else:
                col -= 1
                hasOne = True
        if hasOne:
            return col + 1
        else:
            return -1
            

File: 30-Day LeetCoding Challenge/C++/leftMostColumnWithOne.cpp
/**
 * // This is the BinaryMatrix's API interface.
 * // You should not implement it, or speculate about its implementation
 * class BinaryMatrix {
 *   public:
 *     int get(int row, int col);
 *     vector<int> dimensions();
 * };
 */

class Solution {
public:
    int leftMostColumnWithOne(BinaryMatrix &binaryMatrix) {
        //{rows,cols}
        vector<int> dimensions = binaryMatrix.dimensions();
        int minCol = INT_MAX;
        bool hasOne = false;
        for(int i = 0; i < dimensions[0]; i++) {
            int l = 0;
            int r = dimensions[1] - 1;
            int col = -1;
            while(l <= r) {
                int m = l + (r - l)/2;
                int value = binaryMatrix.get(i, m);
                if(value == 0) {
                    l = m + 1;
                }
                else {
                    r = m - 1;
                    col = m;
                    hasOne = true;
                }
            }
            if(col != -1)
                minCol = min(col, minCol);
        }
        if(hasOne)
            return minCol;
        return -1;
    }
};